File: threads/src/main/java/im/threads/holders/ConsultPhraseHolder.java
package im.threads.holders;

import android.graphics.PorterDuff;
import android.graphics.Typeface;
import android.support.annotation.DrawableRes;
import android.support.annotation.Nullable;
import android.support.v4.content.ContextCompat;
import android.text.TextUtils;
import android.text.format.Formatter;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

import im.threads.R;
import im.threads.formatters.RussianFormatSymbols;
import im.threads.model.ChatStyle;
import im.threads.model.ConsultPhrase;
import im.threads.model.FileDescription;
import im.threads.model.Quote;
import im.threads.opengraph.OGData;
import im.threads.picasso_url_connection_only.Callback;
import im.threads.picasso_url_connection_only.Picasso;
import im.threads.utils.CircleTransform;
import im.threads.utils.FileUtils;
import im.threads.utils.LogUtils;
import im.threads.utils.ViewUtils;
import im.threads.views.CircularProgressButton;

import static android.text.TextUtils.isEmpty;

/**
 * Created by yuri on 08.06.2016.
 * layout/item_consultant_text_with_file.xml
 */
public class ConsultPhraseHolder extends BaseHolder {
    private static final String TAG = "ConsultPhraseHolder ";
    private View fileRow;
    private CircularProgressButton mCircularProgressButton;
    private TextView rightTextHeader;
    private ImageView mImage;
    private TextView mRightTextDescr;
    private TextView rightTextFileStamp;
    private TextView mTimeStampTextView;
    private TextView mPhraseTextView;
    private SimpleDateFormat quoteSdf;
    private SimpleDateFormat timeStampSdf = new SimpleDateFormat("HH:mm");
    public ImageView mConsultAvatar;
    private View mFilterView;
    private View mFilterViewSecond;
    private ChatStyle style;
    private View mBubble;
    private View mPhraseFrame;
    private ViewGroup mOgDataLayout;
    private ImageView mOgImage;
    private TextView mOgTitle;
    private TextView mOgDescription;
    private TextView mOgUrl;
    private TextView mOgTimestamp;
    @DrawableRes
    private int defIcon;

    public ConsultPhraseHolder(ViewGroup parent) {
        super(LayoutInflater.from(parent.getContext()).inflate(R.layout.item_consultant_text_with_file, parent, false));
        fileRow = itemView.findViewById(R.id.right_text_row);
        mCircularProgressButton = itemView.findViewById(R.id.button_download);
        mImage = itemView.findViewById(R.id.image);
        rightTextHeader = itemView.findViewById(R.id.to);
        mRightTextDescr = itemView.findViewById(R.id.file_specs);
        rightTextFileStamp = itemView.findViewById(R.id.send_at);
        mPhraseTextView = itemView.findViewById(R.id.text);
        mConsultAvatar = itemView.findViewById(R.id.consult_avatar);
        mTimeStampTextView = itemView.findViewById(R.id.timestamp);
        mFilterView = itemView.findViewById(R.id.filter);
        mFilterViewSecond = itemView.findViewById(R.id.filter_bottom);
        mPhraseFrame = itemView.findViewById(R.id.phrase_frame);
        mOgDataLayout = itemView.findViewById(R.id.og_data_layout);
        mOgImage = itemView.findViewById(R.id.og_image);
        mOgTitle = itemView.findViewById(R.id.og_title);
        mOgDescription = itemView.findViewById(R.id.og_description);
        mOgUrl = itemView.findViewById(R.id.og_url);
        mOgTimestamp = itemView.findViewById(R.id.og_timestamp);
        if (Locale.getDefault().getLanguage().equalsIgnoreCase("ru")) {
            quoteSdf = new SimpleDateFormat("dd MMMM yyyy", new RussianFormatSymbols());
        } else {
            quoteSdf = new SimpleDateFormat("dd MMMM yyyy");
        }
        mBubble = itemView.findViewById(R.id.bubble);
        if (style == null) style = ChatStyle.getInstance();

        mBubble.setBackground(ContextCompat.getDrawable(itemView.getContext(), style.incomingMessageBubbleBackground));
        mBubble.getBackground().setColorFilter(getColorInt(style.incomingMessageBubbleColor), PorterDuff.Mode.SRC_ATOP);
        setTextColorToViews(new TextView[]{mPhraseTextView,
                rightTextHeader,
                mRightTextDescr,
                rightTextFileStamp}, style.incomingMessageTextColor);

        mTimeStampTextView.setTextColor(getColorInt(style.incomingMessageTimeColor));
        mOgTimestamp.setTextColor(getColorInt(style.incomingMessageTimeColor));

        mPhraseTextView.setLinkTextColor(getColorInt(style.incomingMessageLinkColor));

        defIcon = style.defaultOperatorAvatar;
        setTintToProgressButtonConsult(mCircularProgressButton, style.chatBodyIconsTint);
        itemView.findViewById(R.id.delimeter).setBackgroundColor(getColorInt(style.chatToolbarColorResId));

        mFilterView.setBackgroundColor(ContextCompat.getColor(itemView.getContext(), style.chatHighlightingColor));
        mFilterViewSecond.setBackgroundColor(ContextCompat.getColor(itemView.getContext(), style.chatHighlightingColor));
        mCircularProgressButton.setBackgroundColorResId(style.chatBackgroundColor);

        mConsultAvatar.getLayoutParams().height = (int) itemView.getContext().getResources().getDimension(style.operatorAvatarSize);
        mConsultAvatar.getLayoutParams().width = (int) itemView.getContext().getResources().getDimension(style.operatorAvatarSize);
    }


    public void onBind(ConsultPhrase message, String phrase
            , String avatarPath
            , long timeStamp
            , boolean isAvatarVisible
            , Quote quote
            , FileDescription fileDescription
            , final View.OnClickListener imageClickListener
            , @Nullable View.OnClickListener fileClickListener
            , View.OnLongClickListener onRowLongClickListener
            , View.OnClickListener onAvatarClickListener
            , View.OnClickListener onOgClickListener, boolean isChosen) {

        mConsultAvatar.setImageBitmap(null);
        if (phrase == null) {
            mPhraseTextView.setVisibility(View.GONE);
        } else {
            mPhraseTextView.setVisibility(View.VISIBLE);
            mPhraseTextView.setText(phrase);
        }

        OGData ogData = message.ogData;
        LogUtils.logDev(String.valueOf(ogData));

        if (ogData == null || ogData.isEmpty()) {
            mOgDataLayout.setVisibility(View.GONE);
            mTimeStampTextView.setVisibility(View.VISIBLE);
        } else {
            mOgDataLayout.setVisibility(View.VISIBLE);
            mTimeStampTextView.setVisibility(View.GONE);

            ViewUtils.setClickListener(mOgDataLayout, onOgClickListener);

            if (TextUtils.isEmpty(ogData.title)) {
                mOgTitle.setVisibility(View.GONE);
            } else {
                mOgTitle.setText(ogData.title);
                mOgTitle.setTypeface(mOgTitle.getTypeface(), Typeface.BOLD);
            }

            if (TextUtils.isEmpty(ogData.description)) {
                mOgDescription.setVisibility(View.GONE);
            } else {
                mOgDescription.setText(ogData.description);
            }

            if (TextUtils.isEmpty(ogData.url)) {
                mOgUrl.setVisibility(View.GONE);
            } else {
                mOgUrl.setText(ogData.url);
            }

            if (TextUtils.isEmpty(ogData.image)) {
                mOgImage.setVisibility(View.GONE);
            } else {
                Picasso.with(itemView.getContext())
                        .load(ogData.image)
                        .error(style.imagePlaceholder)
                        .fit()
                        .centerCrop()
                        .into(mOgImage, new Callback() {
                            @Override
                            public void onSuccess() {

                            }

                            @Override
                            public void onError() {
                                mOgImage.setVisibility(View.GONE);
                            }
                        });
            }
        }

        mImage.setVisibility(View.GONE);
        ViewUtils.setClickListener((ViewGroup) itemView, onRowLongClickListener);
        String timeText = timeStampSdf.format(new Date(timeStamp));
        mTimeStampTextView.setText(timeText);
        mOgTimestamp.setText(timeText);
        if (quote != null) {
            fileRow.setVisibility(View.VISIBLE);
            mCircularProgressButton.setVisibility(View.GONE);
            rightTextHeader.setText(quote.getPhraseOwnerTitle());
            mRightTextDescr.setText(quote.getText());
            rightTextFileStamp.setText(itemView.getContext().getString(R.string.threads_sent_at) + " " + quoteSdf.format(new Date(quote.getTimeStamp())));
            if (isEmpty(quote.getPhraseOwnerTitle())) {
                rightTextHeader.setVisibility(View.GONE);
            } else {
                rightTextHeader.setVisibility(View.VISIBLE);
            }
            if (quote.getFileDescription() != null) {
                mCircularProgressButton.setVisibility(View.VISIBLE);
                String filename = quote.getFileDescription().getIncomingName();
                if (filename == null) {
                    filename = FileUtils.getLastPathSegment(quote.getFileDescription().getFilePath()) == null ? "" : FileUtils.getLastPathSegment(quote.getFileDescription().getFilePath());
                }
                mRightTextDescr.setText(filename + "\n" + Formatter.formatFileSize(itemView.getContext(), quote.getFileDescription().getSize()));
                if (fileDescription != null)
                    mCircularProgressButton.setOnClickListener(fileClickListener);
                mCircularProgressButton.setProgress(quote.getFileDescription().getDownloadProgress());
            } else {
                mCircularProgressButton.setVisibility(View.GONE);
            }
        }
        if (fileDescription != null) {
            if (FileUtils.isImage(fileDescription)) {
                fileRow.setVisibility(View.GONE);
                mCircularProgressButton.setVisibility(View.GONE);
                mImage.setVisibility(View.VISIBLE);
                mImage.setOnClickListener(imageClickListener);

                Picasso.with(itemView.getContext())
                        .load(fileDescription.getDownloadPath())
                        .error(style.imagePlaceholder)
                        .fit()
                        .centerCrop()
                        .into(mImage);
            } else {
                fileRow.setVisibility(View.VISIBLE);
                mCircularProgressButton.setVisibility(View.VISIBLE);

                if (fileClickListener != null) {
                    mCircularProgressButton.setOnClickListener(fileClickListener);
                }
                rightTextHeader.setText(fileDescription.getFileSentTo() == null ? "" : fileDescription.getFileSentTo());
                if (!isEmpty(rightTextHeader.getText())) {
                    rightTextHeader.setVisibility(View.VISIBLE);
                } else {
                    rightTextHeader.setVisibility(View.GONE);
                }
                String fileHeader = fileDescription.getIncomingName() == null ? FileUtils.getLastPathSegment(fileDescription.getFilePath()) : fileDescription
                        .getIncomingName();
                mRightTextDescr.setText(fileHeader == null ? "" : fileHeader + "\n" + android.text.format.Formatter
                        .formatFileSize(itemView.getContext(), fileDescription.getSize()));
                rightTextFileStamp
                        .setText(itemView.getContext().getString(R.string.threads_sent_at) + " " + quoteSdf.format(new Date(fileDescription.getTimeStamp())));
                mCircularProgressButton.setProgress(fileDescription.getDownloadProgress());
            }
        }
        if (fileDescription == null && quote == null) {
            fileRow.setVisibility(View.GONE);
        }
        if (fileDescription != null || quote != null) {
            mPhraseFrame.getLayoutParams().width = FrameLayout.LayoutParams.MATCH_PARENT;
        } else {
            mPhraseFrame.getLayoutParams().width = FrameLayout.LayoutParams.WRAP_CONTENT;
        }
        if (isAvatarVisible) {
            float bubbleLeftMarginDp = itemView.getContext().getResources().getDimension(R.dimen.margin_quarter);
            int bubbleLeftMarginPx = ((int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, bubbleLeftMarginDp, itemView.getResources().getDisplayMetrics()));
            RelativeLayout.LayoutParams lp = (RelativeLayout.LayoutParams) mBubble.getLayoutParams();
            lp.setMargins(bubbleLeftMarginPx, lp.topMargin, lp.rightMargin, lp.bottomMargin);
            mBubble.setLayoutParams(lp);

            mConsultAvatar.setVisibility(View.VISIBLE);
            mConsultAvatar.setOnClickListener(onAvatarClickListener);
            if (!isEmpty(avatarPath)) {
                avatarPath = FileUtils.convertRelativeUrlToAbsolute(itemView.getContext(), avatarPath);
                Picasso
                        .with(itemView.getContext())
                        .load(avatarPath)
                        .fit()
                        .noPlaceholder()
                        .centerCrop()
                        .transform(new CircleTransform())
                        .into(mConsultAvatar, new Callback() {
                            @Override
                            public void onSuccess() {
                            }

                            @Override
                            public void onError() {
                                showDefIcon();
                            }
                        });
            } else {
                showDefIcon();
            }
        } else {
            mConsultAvatar.setVisibility(View.GONE);

            int avatarSizeRes = style.operatorAvatarSize;
            int avatarSizePx = itemView.getContext().getResources().getDimensionPixelSize(avatarSizeRes);

            int bubbleLeftMarginPx = itemView.getContext().getResources().getDimensionPixelSize(R.dimen.margin_half);
            int avatarLeftMarginPx = itemView.getContext().getResources().getDimensionPixelSize(R.dimen.margin_half);

            RelativeLayout.LayoutParams lp = (RelativeLayout.LayoutParams) mBubble.getLayoutParams();
            lp.setMargins(avatarSizePx + bubbleLeftMarginPx + avatarLeftMarginPx, lp.topMargin, lp.rightMargin, lp.bottomMargin);
            mBubble.setLayoutParams(lp);

        }

        mFilterView.setVisibility(isChosen ? View.VISIBLE : View.INVISIBLE);
        mFilterViewSecond.setVisibility(isChosen && isAvatarVisible ? View.VISIBLE : View.INVISIBLE);
    }

    private void showDefIcon() {
        Picasso.with(itemView.getContext())
                .load(defIcon)
                .centerInside()
                .noPlaceholder()
                .fit()
                .transform(new CircleTransform())
                .into(mConsultAvatar);
    }
}

File: threads/src/main/java/im/threads/ui/holders/ConsultPhraseHolder.kt
package im.threads.ui.holders

import android.app.ActionBar
import android.text.TextUtils
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import android.view.View.OnLongClickListener
import android.view.ViewGroup
import android.view.ViewGroup.MarginLayoutParams
import android.view.animation.Animation
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.appcompat.content.res.AppCompatResources
import androidx.core.content.ContextCompat
import androidx.core.graphics.BlendModeColorFilterCompat
import androidx.core.graphics.BlendModeCompat
import androidx.core.view.isVisible
import im.threads.R
import im.threads.business.imageLoading.ImageLoader
import im.threads.business.imageLoading.ImageModifications
import im.threads.business.models.ChatItem
import im.threads.business.models.ConsultPhrase
import im.threads.business.models.FileDescription
import im.threads.business.models.enums.AttachmentStateEnum
import im.threads.business.models.enums.ModificationStateEnum
import im.threads.business.ogParser.OGDataContent
import im.threads.business.ogParser.OpenGraphParser
import im.threads.business.utils.FileUtils.isImage
import im.threads.business.utils.UrlUtils
import im.threads.ui.config.Config
import im.threads.ui.utils.ColorsHelper
import im.threads.ui.utils.gone
import im.threads.ui.utils.visible
import im.threads.ui.views.CircularProgressButton
import im.threads.ui.widget.textView.BubbleMessageTextView
import im.threads.ui.widget.textView.BubbleTimeTextView
import io.reactivex.subjects.PublishSubject
import java.lang.ref.WeakReference
import java.util.Date

/** layout/item_consultant_text_with_file.xml */
class ConsultPhraseHolder(
    private val parent: ViewGroup,
    private val maskedTransformation: ImageModifications.MaskedModification?,
    highlightingStream: PublishSubject<ChatItem>,
    openGraphParser: OpenGraphParser
) : BaseHolder(
    LayoutInflater.from(parent.context)
        .inflate(R.layout.ecc_item_consultant_text_with_file, parent, false),
    highlightingStream,
    openGraphParser
) {
    private val fileRow: View = itemView.findViewById(R.id.rightTextRow)
    private val quoteLayout: LinearLayout = itemView.findViewById(R.id.quoteLayout)
    private val quoteTextHeader: TextView = itemView.findViewById(R.id.quoteTo)
    private val quoteTextDescription: TextView = itemView.findViewById(R.id.quoteFileSpecs)
    private val quoteTextTimeStamp: TextView = itemView.findViewById(R.id.quoteSendAt)
    private val quoteFileImage = itemView.findViewById<ImageView>(R.id.quoteFileImage)
    private val quoteProgressButton: CircularProgressButton = itemView.findViewById(R.id.quoteButtonDownload)
    private val circularProgressButton: CircularProgressButton = itemView.findViewById(R.id.buttonDownload)
    private val errorTextView: TextView = itemView.findViewById(R.id.errorText)
    private val fileImage = itemView.findViewById<ImageView>(R.id.fileImage)
    private val rightTextHeader: TextView = itemView.findViewById(R.id.to)
    private val image: ImageView = itemView.findViewById(R.id.image)
    private val imageRoot: FrameLayout = itemView.findViewById(R.id.imageRoot)
    private val imageLayout: FrameLayout = itemView.findViewById(R.id.imageLayout)
    private val errorImage: ImageView = itemView.findViewById(R.id.errorImage)
    private val loaderImage: ImageView = itemView.findViewById(R.id.loaderImage)
    private val rightTextDescription: TextView = itemView.findViewById(R.id.fileSpecs)
    private val rightTextFileStamp: TextView = itemView.findViewById(R.id.sendAt)
    private val timeStampTextView =
        itemView.findViewById<BubbleTimeTextView>(R.id.timeStamp).apply {
            setTextColor(getColorInt(style.incomingMessageTimeColor))
            if (style.incomingMessageTimeTextSize > 0) {
                setTextSize(
                    TypedValue.COMPLEX_UNIT_PX,
                    itemView.context.resources.getDimension(style.incomingMessageTimeTextSize)
                )
            }
        }
    private val phraseTextView = itemView.findViewById<BubbleMessageTextView>(R.id.text).apply {
        setLinkTextColor(getColorInt(style.incomingMessageLinkColor))
    }
    private val consultAvatar = itemView.findViewById<ImageView>(R.id.consultAvatar).apply {
        layoutParams.height =
            itemView.context.resources.getDimension(style.operatorAvatarSize)
                .toInt()
        layoutParams.width =
            itemView.context.resources.getDimension(style.operatorAvatarSize)
                .toInt()
    }

    private val phraseFrame: View = itemView.findViewById(R.id.phraseFrame)
    private val ogDataLayout: ViewGroup = itemView.findViewById(R.id.ogDataLayout)
    private val ogTimestamp = itemView.findViewById<BubbleTimeTextView>(R.id.ogTimeStamp).apply {
        setTextColor(getColorInt(style.incomingMessageTimeColor))
    }
    private val bubbleLayout = itemView.findViewById<ViewGroup>(R.id.bubble).apply {
        background =
            AppCompatResources.getDrawable(
                itemView.context,
                style.incomingMessageBubbleBackground
            )

        background.colorFilter =
            BlendModeColorFilterCompat.createBlendModeColorFilterCompat(
                getColorInt(style.incomingMessageBubbleColor),
                BlendModeCompat.SRC_ATOP
            )
    }

    init {
        itemView.findViewById<View>(R.id.delimiter)
            .setBackgroundColor(getColorInt(style.incomingDelimitersColor))
        itemView.findViewById<View>(R.id.quoteDelimiter)
            .setBackgroundColor(getColorInt(style.incomingDelimitersColor))
        setTextColorToViews(
            arrayOf(
                phraseTextView,
                rightTextHeader,
                rightTextDescription,
                rightTextFileStamp
            ),
            style.incomingMessageTextColor
        )
        ColorsHelper.setTextColor(errorTextView, style.errorMessageTextColor)
        setUpProgressButton(circularProgressButton)
    }

    fun onBind(
        consultPhrase: ConsultPhrase,
        highlighted: Boolean,
        imageClickListener: View.OnClickListener,
        fileClickListener: View.OnClickListener,
        onQuoteClickListener: View.OnClickListener,
        onRowLongClickListener: OnLongClickListener,
        onAvatarClickListener: View.OnClickListener
    ) {
        setupPaddingsAndBorders(consultPhrase.fileDescription)
        subscribeForHighlighting(consultPhrase, itemView)
        subscribeForOpenGraphData(
            OGDataContent(
                WeakReference(ogDataLayout),
                WeakReference(ogTimestamp),
                WeakReference(timeStampTextView),
                consultPhrase.phraseText
            )
        )
        viewUtils.setClickListener(itemView as ViewGroup, onRowLongClickListener)
        showConsultTimeStamp(consultPhrase, listOf(ogTimestamp, timeStampTextView))
        imageLayout.isVisible = false
        fileRow.isVisible = false
        quoteLayout.isVisible = false
        imageRoot.isVisible = false
        setLayoutMargins(true, bubbleLayout)
        showAvatar(consultAvatar, consultPhrase, onAvatarClickListener)
        changeHighlighting(highlighted)

        if (consultPhrase.modified == ModificationStateEnum.DELETED) {
            phraseTextView.setTextSize(TypedValue.COMPLEX_UNIT_PX, itemView.getContext().resources.getDimension(style.systemMessageTextSize))
            phraseTextView.setTextColor(ContextCompat.getColor(itemView.getContext(), style.systemMessageTextColorResId))
            phraseTextView.bindTimestampView(timeStampTextView)
            phraseTextView.visible()
            phraseTextView.text = this.parent.context.getString(R.string.ecc_message_deleted)
        } else {
            setTextColorToViews(arrayOf(phraseTextView), style.incomingMessageTextColor)
            phraseTextView.setTextSize(TypedValue.COMPLEX_UNIT_PX, itemView.getContext().resources.getDimension(R.dimen.ecc_text_medium))
            phraseTextView.text = ""
            consultPhrase.phraseText?.let {
                showPhrase(consultPhrase, it.trim())
            } ?: run {
                phraseTextView.isVisible = false
            }

            consultPhrase.quote?.let {
                showQuote(
                    it,
                    onQuoteClickListener,
                    quoteLayout,
                    quoteTextHeader,
                    quoteTextDescription,
                    quoteTextTimeStamp,
                    quoteFileImage,
                    quoteProgressButton
                )
            } ?: run {
                quoteLayout.isVisible = false
            }

            consultPhrase.fileDescription?.let {
                when (it.state) {
                    AttachmentStateEnum.PENDING -> {
                        showLoaderLayout(it)
                    }
                    AttachmentStateEnum.ERROR -> {
                        showErrorLayout(it)
                    }
                    else -> {
                        showCommonLayout(it, fileClickListener, imageClickListener)
                    }
                }
            } ?: run {
                consultPhrase.formattedPhrase?.let {
                    UrlUtils.extractImageMarkdownLink(it)?.let { imageUrl ->
                        loadImage(imageUrl, imageClickListener, true)
                    }
                }
            }
            if (consultPhrase.fileDescription != null || consultPhrase.quote != null) {
                phraseFrame.layoutParams.width = FrameLayout.LayoutParams.MATCH_PARENT
            } else {
                phraseFrame.layoutParams.width = FrameLayout.LayoutParams.WRAP_CONTENT
            }
        }
    }

    private fun setupPaddingsAndBorders(fileDescription: FileDescription?) = with(bubbleLayout) {
        val chatStyle = Config.getInstance().chatStyle
        val resources = context.resources
        val borderLeft = resources.getDimensionPixelSize(chatStyle.incomingImageLeftBorderSize)
        val borderTop = resources.getDimensionPixelSize(chatStyle.incomingImageTopBorderSize)
        val borderRight = resources.getDimensionPixelSize(chatStyle.incomingImageRightBorderSize)
        val borderBottom = resources.getDimensionPixelSize(chatStyle.incomingImageBottomBorderSize)
        val isBordersNotSet = borderLeft == 0 && borderTop == 0 && borderRight == 0 && borderBottom == 0
        val isImage = isImage(fileDescription)

        setImageSize(imageRoot)
        setLayoutMargins(true, bubbleLayout)
        if (isImage && fileDescription?.state == AttachmentStateEnum.READY) {
            imageRoot.visible()

            (bubbleLayout.layoutParams as MarginLayoutParams).let {
                it.marginEnd = 0
                bubbleLayout.layoutParams = it
            }
            bubbleLayout.invalidate()
            bubbleLayout.requestLayout()

            if (isBordersNotSet) {
                phraseFrame.setPadding(0, 0, 0, 0)
                quoteLayout.setPadding(0, 0, 0, 0)
                setPaddings(true, this)
            } else {
                setPadding(0, 0, 0, 0)
                (image.layoutParams as FrameLayout.LayoutParams).apply {
                    setMargins(borderLeft, borderTop, borderRight, borderBottom)
                    image.layoutParams = this
                }
                phraseFrame.setPadding(
                    borderLeft,
                    0,
                    borderRight,
                    resources.getDimensionPixelSize(style.bubbleIncomingPaddingBottom)
                )
                quoteLayout.setPadding(
                    borderLeft,
                    resources.getDimensionPixelSize(style.bubbleIncomingPaddingTop),
                    borderRight,
                    0
                )
            }
            image.invalidate()
            image.requestLayout()
        } else {
            (bubbleLayout.layoutParams as MarginLayoutParams).let {
                it.width = ActionBar.LayoutParams.WRAP_CONTENT
                it.height = ActionBar.LayoutParams.WRAP_CONTENT
                it.marginEnd = resources.getDimensionPixelSize(R.dimen.ecc_user_margin_right)
            }
            bubbleLayout.invalidate()
            bubbleLayout.requestLayout()

            imageRoot.gone()

            phraseFrame.setPadding(0, 0, 0, 0)
            quoteLayout.setPadding(0, 0, 0, 0)
            setPaddings(true, this)
        }
    }

    private fun loadImage(
        fileDescription: FileDescription,
        imageClickListener: View.OnClickListener,
        isExternalImage: Boolean = false
    ) {
        fileDescription.getPreviewFileDescription()?.let {
            val previewUri = getPreviewUri(it.getPreviewFileDescription())
            val fileUri = if (previewUri?.toString().isNullOrBlank()) {
                it.downloadPath
            } else {
                previewUri?.toString()
            }
            loadImage(fileUri, imageClickListener, isExternalImage)
        }
    }

    private fun loadImage(
        imagePath: String?,
        imageClickListener: View.OnClickListener,
        isExternalImage: Boolean = false
    ) {
        hideErrorImage(imageLayout, errorImage)
        fileRow.gone()
        circularProgressButton.gone()
        image.visible()
        imageRoot.visible()
        image.setOnClickListener(imageClickListener)
        if (!imagePath.isNullOrEmpty()) {
            startLoaderAnimation()
            val loadConfig = ImageLoader
                .get()
                .load(imagePath)
                .autoRotateWithExif(true)
                .scales(ImageView.ScaleType.FIT_XY, ImageView.ScaleType.CENTER_CROP)
                .modifications(maskedTransformation)
                .callback(object : ImageLoader.ImageLoaderCallback {
                    override fun onImageLoaded() {
                        stopLoaderAnimation()
                    }

                    override fun onImageLoadError() {
                        showErrorImage(imageLayout, errorImage)
                        stopLoaderAnimation()
                    }
                })
            if (isExternalImage) {
                loadConfig.disableEdnaSsl()
            }
            loadConfig.into(image)
        } else {
            image.setImageResource(style.imagePlaceholder)
        }
    }

    private fun showPhrase(
        consultPhrase: ConsultPhrase,
        phrase: String
    ) {
        phraseTextView.bindTimestampView(timeStampTextView)
        phraseTextView.visible()
        val deeplink = UrlUtils.extractDeepLink(phrase)
        val extractedLink = if (style.linkPreviewEnabled) bindOGData(phrase) else UrlUtils.extractLink(phrase)
        val emails = UrlUtils.extractEmailAddresses(phrase)
        highlightOperatorText(
            phraseTextView,
            consultPhrase.formattedPhrase,
            consultPhrase.phraseText,
            deeplink ?: extractedLink?.link,
            emails
        )
    }

    private fun startLoaderAnimation() {
        imageLayout.visibility = View.VISIBLE
        loaderImage.visibility = View.VISIBLE
        image.visibility = View.INVISIBLE
        initAnimation(loaderImage, true)
        rotateAnim.duration = 3000
        rotateAnim.repeatCount = Animation.INFINITE
        loaderImage.animation = rotateAnim
        rotateAnim.start()
    }

    private fun stopLoaderAnimation() {
        loaderImage.visibility = View.INVISIBLE
        image.visibility = View.VISIBLE
        rotateAnim.cancel()
        rotateAnim.reset()
    }

    private fun showLoaderLayout(fileDescription: FileDescription) {
        fileRow.isVisible = true
        fileImage.isVisible = true
        errorTextView.isVisible = false
        rightTextDescription.text = fileDescription.incomingName
        circularProgressButton.isVisible = false
        imageRoot.isVisible = false
        fileImage.background = null
        initAnimation(fileImage, true)
    }

    private fun showErrorLayout(fileDescription: FileDescription) {
        fileRow.isVisible = true
        fileImage.isVisible = true
        errorTextView.isVisible = true
        imageLayout.isVisible = false
        circularProgressButton.isVisible = false
        fileImage.background = null
        imageRoot.isVisible = false
        fileImage.setImageResource(getErrorImageResByErrorCode(fileDescription.errorCode))
        rightTextDescription.text = fileDescription.incomingName
        val errorString = getString(getErrorStringResByErrorCode(fileDescription.errorCode))
        errorTextView.text = errorString
        rotateAnim.cancel()
    }

    private fun showCommonLayout(
        fileDescription: FileDescription,
        fileClickListener: View.OnClickListener,
        imageClickListener: View.OnClickListener
    ) {
        imageRoot.isVisible = false
        fileRow.isVisible = true
        errorTextView.isVisible = false
        circularProgressButton.isVisible = true
        rotateAnim.cancel()
        val isStateReady = fileDescription.state == AttachmentStateEnum.READY
        if (isStateReady && isImage(fileDescription)) {
            loadImage(fileDescription, imageClickListener)
        } else if (!isStateReady && isImage(fileDescription)) {
            startLoaderAnimation()
        } else {
            fileRow.visibility = View.VISIBLE
            circularProgressButton.isVisible = true
            circularProgressButton.isClickable = true
            circularProgressButton.setOnClickListener(fileClickListener)
            rightTextHeader.text =
                if (fileDescription.from == null) "" else fileDescription.from
            if (!TextUtils.isEmpty(rightTextHeader.text)) {
                rightTextHeader.visibility = View.VISIBLE
            } else {
                rightTextHeader.visibility = View.GONE
            }
            fileNameFromDescription(fileDescription) { fileName ->
                getFileDescriptionText(fileName, fileDescription)
            }
            rightTextFileStamp.text = itemView.context.getString(
                R.string.ecc_sent_at,
                quoteSdf.format(Date(fileDescription.timeStamp))
            )
            circularProgressButton.setProgress(if (fileDescription.fileUri != null) 100 else fileDescription.downloadProgress)
        }
    }
}
